Backend PanelApp implementation for #1891

tommyli opened the pull request:
Backend implementation of PanelApp ready for your team to review (#1891).  In particular:

* Is using Django apps + ENV var a good approach for keeping such features optional?  If your team is happy with it then it may be a good model going forward to collaborate new features on?
* Another approach is to implement it similar to other `reference_data_*` but I feel current impl keeps panelapp even more separate. 
* The UI is not implemented but the changes to the store are what's needed for the UI store to hold new info for PanelApp fields.  These changes should be safe regardless if `panelapp` is installed in the backend.

tommyli commented:
I'm having trouble understanding why the Codacy docstring issues appear when I'm following the same convention as other docstrings in the codebase.

hanars commented on ui/redux/selectors.js:
```diff
@@ -20,6 +20,7 @@ export const getVariantFunctionalDataByGuid = state => state.variantFunctionalDa
 export const getMmeSubmissionsByGuid = state => state.mmeSubmissionsByGuid
 export const getMmeResultsByGuid = state => state.mmeResultsByGuid
 export const getGenesById = state => state.genesById
+export const getPaGenesById = state => state.pagenesById
```
you don't seem to use this data anywhere, or the `paLocusList` data on the locus list itself. Is there a separate PR where you show this information? If not, it would be way more performant not to fetch/ return the data to the client

tommyli replied:
See answer below regarding UI.

hanars commented:
> I'm having trouble understanding why the Codacy docstring issues appear when I'm following the same convention as other docstrings in the codebase.

because like 99% of the docstrings in seqr fail codacy, we've never gone back and fixed legacy issues

hanars commented on panelapp/panelapp_utils.py:
```diff
@@ -0,0 +1,146 @@
+    else:
+        seqr_locus_list = create_model_from_json(SeqrLocusList, {
+            'name': panel_json['name'],
+            'description': None,
```
It probably would be clearer for users if the description indicated that this was imported from panel app. Either that or wherever you show the description in the UI, show some additional panel app data to make it clearer

tommyli replied:
Yes agree, it'll be more consistent with requests like [Require descriptions for new gene lists](https://github.com/broadinstitute/seqr/issues/1983).  There are a few fields from `PaLocusList` I could use to populate the description but we have yet to agree on how PanelApp gene lists should be displayed.  See answer below regarding UI.

hanars commented:
> * Is using Django apps + ENV var a good approach for keeping such features optional?  If your team is happy with it then it may be a good model going forward to collaborate new features on?

I think its a pretty clean approach to conditionally add functionality, and for future feature requests I think it would be a good approach

> * The UI is not implemented but the changes to the store are what's needed for the UI store to hold new info for PanelApp fields.  These changes should be safe regardless if `panelapp` is installed in the backend.

The changes to the UI are safe as is, but not very useful. Based on this branch, if there are any UI enhancements that come with the panel app no one using this repo would be able to see them. I think it would be better to include the code changes for the UI display and either 
- set a constant in the metadata that is used by the client code to conditionally show features, similar to `googleLoginEnabled` OR
- make sure the client code gracefully handles the case where the store variables are empty/ null and don't change the display in that case. 

However, after talking to the team we are actually going to take this live for our group! So we are happy to take this code as-is and set the neccessary environment variable to enable it, or if you would prefer to take out the conditionals we are happy to take it as an unconditional part of the seqr code base as well

tommyli commented on .github/workflows/unit-tests.yml:
```diff
@@ -55,6 +55,12 @@ jobs:
+        ENABLE_PANEL_APP: "True"
+      run: |
+        coverage run --source="./panelapp" --omit="*/migrations/*","*/apps.py" manage.py test -p '*_tests.py' -v 2 panelapp
+        coverage report --fail-under=99
```
Yeah, I initially tried something like `@mock.patch` but it didn't work because the tests required `panelapp` DB migrations.  There are ways to mock in `settings.installed_apps` too but it made the test a bit cumbersome so I thought this was a cleaner approach.

tommyli commented on panelapp/views.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@data_manager_required
+def import_panelapp_handler(request):
```
Fair enough.  Our instance can only be accessed via VPN so we're a little spoilt in the security sense.

We were going to implement a UI trigger for this but your idea of using a cron job on deployed host is probably even better and safer.  Will remove this and keep the manage command.

tommyli commented on seqr/views/utils/orm_to_json_utils.py:
```diff
@@ -722,9 +728,17 @@ def _process_result(result, locus_list):
     prefetch_related_objects(locus_lists, 'locuslistgene_set')
     prefetch_related_objects(locus_lists, 'locuslistinterval_set')
+    if ENABLE_PANEL_APP:
+        prefetch_related_objects(locus_lists, 'palocuslist')
```
Will fix.

tommyli commented:
> However, after talking to the team we are actually going to take this live for our group! So we are happy to take this code as-is and set the neccessary environment variable to enable it, or if you would prefer to take out the conditionals we are happy to take it as an unconditional part of the seqr code base as well

Great!  I will append to this PR with fixes that addresses some of the above issues and remove the `ENABLE_PANEL_APP` conditionals.  That'll make the codebase even cleaner.  The history of this PR will retain the history on an approach to feature toggle.

tommyli commented:
> The changes to the UI are safe as is, but not very useful. Based on this branch, if there are any UI enhancements that come with the panel app no one using this repo would be able to see them. I think it would be better to include the code changes for the UI display and either

Indeed, a separate PR will be raised for the UI changes.  Here are some of our thoughts:
* I was so happy to see your team implemented [Gene list search](https://github.com/broadinstitute/seqr/pull/2051) :).  It's great timing so whatever we do, it'll be built on top of that.  From the backend changes of #2051, I may end up raising a new PR that supersedes this current one and raise another PR for the UI.
* Whatever the UI design, we will ensure the UI will handle cases whether PanelApp gene lists exists or not.
* The additional fields from PanelApp included in `PaLocusList` do not really "fit" into the current UI table.  We may end up having a third section (i.e. "My Gene Lists" + "Public Gene Lists" + "PanelApp").  However, the users are happy to simply have those fields crammed into `description`, that'll be good.
* The Gene Lists details page will also be updated to include additional data such as `confidence_level`.

Future changes, most likely separate PRs.
* PanelApp genes also includes data such as "mode of inheritance" and "disease phenotypes".  These will be incorporated in variant search results.
* Add new search/filter in the gene lists filter when doing variant search

Any comments/thoughts/ideas welcome of course.

hanars commented:
One comment on separate PRs is we don't like to merge incomplete functionality. So if we want to keep the front end code to a separate PR thats fine, but then this PR shouldn't have any front-end changes in it, including changes to the data returned to the client. right now if we released this code we would be slowing down a bunch of API requests to return data we wouldn't use, which isn't great. So I would say wither we want this PR to include some of the basic UI enhancements, maybe populating the "description" column with structured data from panel app if it exists, or we should seperate all that functionality into the future UI PR

tommyli commented:
Closing, please refer to new PR #2107 which supersedes this.
